### PR TITLE
Make a generic OIDC provider

### LINKED ISSUE #2501: Add support for Auth0
We need to add support for mozilla auth0 for bloomington

### PR DESCRIPTION
**Pull request by kmcnellis**
### Description
Enable a generic, fully-configurable OIDC provider for applicant authentication

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #2501

### CODE REVIEW

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/auth/oidc/applicant/GenericOidcProvider.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  @Override
+  public ProfileCreator getProfileAdapter(OidcConfiguration config, OidcClient client) {
+    Optional<String> firstNameAttr = getConfigurationValue(FIRST_NAME_ATTRIBUTE_CONFIG_NAME);
+    Optional<String> secondNameAttr = getConfigurationValue(SECOND_NAME_ATTRIBUTE_CONFIG_NAME);
+    String emailAttr = getConfigurationValueOrThrow(EMAIL_ATTRIBUTE_CONFIG_NAME);
+    Optional<String> localeAttr = getConfigurationValue(LOCALE_ATTRIBUTE_CONFIG_NAME);
+
+    var nameAttrsBuilder = ImmutableList.<String>builder();
+    firstNameAttr.ifPresent(n -> nameAttrsBuilder.add(n));
+    secondNameAttr.ifPresent(n -> nameAttrsBuilder.add(n));
```
[the remark this answers is not in the document]
Comment (kmcnellis) [reply server/app/auth/oidc/applicant/GenericOidcProvider.java]: done

File: server/conf/application.conf [hunk 2]
```diff
@@ -88,6 +88,32 @@ login_radius.keystore_name = ${?LOGIN_RADIUS_KEYSTORE_NAME}
...
+generic_oidc.email_attribute = "user_emailid"
+generic_oidc.email_attribute = ${?OIDC_EMAIL_ATTRIBUTE}
+
+# first_name_attribute, middle_name_attribute, and last_name_attribute are
+# concatenated if set and present in the oidc response.
+generic_oidc.first_name_attribute ="user_displayname"
+generic_oidc.first_name_attribute = ${?OIDC_FIRST_NAME_ATTRIBUTE}
+generic_oidc.middle_name_attribute = ${?OIDC_MIDDLE_NAME_ATTRIBUTE}
+generic_oidc.last_name_attribute = ${?OIDC_LAST_NAME_ATTRIBUTE}
+
```
[the remark this answers is not in the document]
Comment (kmcnellis) [reply server/conf/application.conf]: Renamed the convig values, but the GenericOidcProfileAdapter is already inside the applicant package.  It seems a little unnessesary to further clarify that.

[the remark this answers is not in the document]
Comment (bion) [reply server/conf/application.conf]: sgtm!

File: server/app/auth/oidc/applicant/GenericOidcProvider.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      Provider<UserRepository> applicantRepositoryProvider) {
+    super(configuration, profileFactory, applicantRepositoryProvider);
+  }
+
+  @Override
+  @VisibleForTesting
+  public String attributePrefix() {
+    return ATTRIBUTE_PREFIX;
+  }
+  ;
```
Comment (bion) [inline server/app/auth/oidc/applicant/GenericOidcProvider.java:49]:
```suggestion
```

**Review by bion — CHANGES_REQUESTED** [review]

**Review by azizova-leyla — APPROVED** [review]

File: server/app/auth/oidc/applicant/GenericOidcProvider.java (hunk 3 above)
Comment (kmcnellis) [reply server/app/auth/oidc/applicant/GenericOidcProvider.java]: done

**Review by bion — APPROVED** [review]

**Pull request merged**